File: ruby/mortgage.rb
#coding: utf-8
require 'optparse'

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: mortage.rb [options]"

  opts.on("-a", "--amount [AMOUNT]", "Amount borrowed") do |a| 
     options[:amount] = a.to_i
  end
  opts.on("-r", "--rate [RATE]", "Interest rate") do |r| 
     options[:rate] = r.to_f
  end
  opts.on("-d", "--duration [DURATION]", "Loan duration in years") do |d| 
     options[:duration] = d.to_i
  end
end.parse!

v_rm = options[:rate] / 12
v_dm = options[:duration].to_i * 12
v_amount = options[:amount].to_f

top_line = v_amount * v_rm * (( 1 + (v_rm / 100))**v_dm) 
bottom_line = 100 * (( (1 + ( v_rm / 100 ))**v_dm) -1)
mp = top_line / bottom_line
ap = mp * 12
printf "The amount borrowed is £%0.2f\n", v_amount.round(2)
printf "The interest rate is %s\%\n", options[:rate].to_s
printf "The loan duration is %s years\n", options[:duration].to_s
printf "Annual repayments will be £%0.2f\n", ap.round(2)
printf "Monthly repayments will be £%0.2f\n", mp.round(2)
printf "Total cost of loan will be £%0.2f\n", (ap * options[:duration]).round(2)

File: python3/mortgage.py
# -*- coding: utf-8 -*-
from optparse import OptionParser
import math
def main () :
    parser = OptionParser()
    parser.add_option("-a", "--amount", dest="amount",
                      help="the amount to be borrowed", metavar="AMOUNT")
    parser.add_option("-i", "--interest", dest="interest",
                      help="the annual interest rate", metavar="INTEREST")
    parser.add_option("-d", "--duration", dest="duration",
                      help="Duration of the loan, in years", metavar="DURATION")
    (options, args) = parser.parse_args()
    if not options.amount :
       print("You must specify an amount")
       exit(-1)
    if not options.interest :
       print("You must specify an interest rate")
       exit(-1)
    if not options.duration :
       print("You must specify an duration")
       exit(-1)
    v_amount = int(options.amount)
    v_interest = float(options.interest)
    v_duration = int(options.duration)
    monthly_interest = float(v_interest) / 12
    duration_in_months = v_duration * 12
    top_line = ( v_amount * monthly_interest * ( 1 + ( monthly_interest / 100 ))**duration_in_months );
    bottom_line = 100 * (( (1 + ( monthly_interest / 100 ))**duration_in_months) -1);
    monthly_payment = top_line / bottom_line ;
    annual_payment = monthly_payment * 12;
    print("The amount borrowed is " +  chr(163) + "%0.2f" % (v_amount))
    print("The interest rate is {:.2f}".format(v_interest))
    print("The loan duration is {:0>2d} years".format(v_duration))
    print("Annual replayments will be " +  chr(163) + "%0.2f" % (annual_payment,))
    print("Monthly replayments will be " +  chr(163) + "%0.2f" % (monthly_payment,))
    print("Total cost of loan will be " +  chr(163) + "%0.2f" % (annual_payment * v_duration))

main()
